File: src/models/admin_knowledge_base.py
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field

# Public API fields intentionally follow the frontend camelCase contract.
# ruff: noqa: N815


DocumentStatus = Literal["ACTIVE", "INACTIVE", "ARCHIVED"]

DocumentType = Literal[
    "PDF",
    "DOCX",
]

RagDocumentType = Literal[
    "policy",
    "form",
    "agreement",
    "talent_handbook",
    "capstone_booklet",
    "knowledge",
]


class AdminKnowledgeUser(BaseModel):
    id: int
    fullName: str
    email: str


class AdminKnowledgeDocumentVersion(BaseModel):
    id: int
    version: str
    fileUrl: str | None = None
    fileHash: str | None = None
    extractedTextPath: str | None = None
    chunkPath: str | None = None
    effectiveDate: str | None = None
    status: str
    createdAt: str | None = None


class AdminKnowledgeIndexJob(BaseModel):
    id: int
    jobType: str
    status: str
    chunksCreated: int = 0
    errorMessage: str | None = None
    startedAt: str | None = None
    completedAt: str | None = None
    createdAt: str | None = None


class AdminKnowledgeDocumentListItem(BaseModel):
    id: int
    title: str
    documentType: DocumentType
    ragDocumentType: RagDocumentType | None = None
    description: str | None = None
    fileUrl: str | None = None
    currentVersion: str | None = None
    year: int | None = None
    status: str
    uploadedBy: AdminKnowledgeUser | None = None
    createdAt: str | None = None
    updatedAt: str | None = None
    currentVersionInfo: AdminKnowledgeDocumentVersion | None = None
    latestIndexJob: AdminKnowledgeIndexJob | None = None


class AdminKnowledgeDocumentFilters(BaseModel):
    types: list[str] = Field(default_factory=list)
    statuses: list[str] = Field(default_factory=list)
    years: list[int] = Field(default_factory=list)


class AdminKnowledgeDocumentsResponse(BaseModel):
    items: list[AdminKnowledgeDocumentListItem] = Field(default_factory=list)
    total: int = 0
    page: int = 1
    pageSize: int = 10
    totalPages: int = 0
    filters: AdminKnowledgeDocumentFilters = Field(
        default_factory=AdminKnowledgeDocumentFilters,
    )


class AdminKnowledgeDocumentDetail(AdminKnowledgeDocumentListItem):
    versions: list[AdminKnowledgeDocumentVersion] = Field(default_factory=list)


class AdminKnowledgeDocumentDetailResponse(BaseModel):
    document: AdminKnowledgeDocumentDetail


class AdminKnowledgeDocumentCreateRequest(BaseModel):
    title: str = Field(min_length=1, max_length=255)
    documentType: DocumentType
    ragDocumentType: RagDocumentType
    description: str | None = None
    fileUrl: str | None = None
    currentVersion: str | None = Field(default=None, max_length=30)
    year: int | None = None
    status: DocumentStatus = "ACTIVE"


class AdminKnowledgeDocumentUpdateRequest(BaseModel):
    title: str | None = Field(default=None, min_length=1, max_length=255)
    documentType: DocumentType | None = None
    ragDocumentType: RagDocumentType | None = None
    description: str | None = None
    fileUrl: str | None = None
    currentVersion: str | None = Field(default=None, max_length=30)
    year: int | None = None
    status: DocumentStatus | None = None


class AdminKnowledgeDocumentActionResponse(BaseModel):
    documentId: int
    message: str


class AdminKnowledgeDocumentVersionsResponse(BaseModel):
    items: list[AdminKnowledgeDocumentVersion] = Field(default_factory=list)


class AdminKnowledgeVersionActionResponse(BaseModel):
    documentId: int
    versionId: int
    message: str


class AdminRagChunkSummary(BaseModel):
    total: int = 0
    documents: int = 0
    translated: int = 0
    averageCharacters: int = 0


class AdminRagChunkFilters(BaseModel):
    documentNames: list[str] = Field(default_factory=list)
    documentTypes: list[str] = Field(default_factory=list)
    languages: list[str] = Field(default_factory=list)


class AdminRagChunkListItem(BaseModel):
    chunkId: str
    position: int
    documentName: str
    documentType: str
    sourcePriority: int
    contentPreview: str
    language: str
    page: int | None = None
    section: str | None = None
    subsection: str | None = None
    topic: str | None = None
    policyVersion: str | None = None
    effectiveDate: str | None = None
    ingestedAt: str | None = None
    characterCount: int = 0
    wordCount: int = 0
    sourceElementCount: int = 0
    hasTranslation: bool = False


class AdminRagChunksResponse(BaseModel):
    items: list[AdminRagChunkListItem] = Field(default_factory=list)
    summary: AdminRagChunkSummary = Field(default_factory=AdminRagChunkSummary)
    filters: AdminRagChunkFilters = Field(default_factory=AdminRagChunkFilters)
    activeBuildId: str
    page: int = 1
    pageSize: int = 25
    total: int = 0
    totalPages: int = 0


class AdminRagChunkDetail(AdminRagChunkListItem):
    contentOriginal: str
    contentVi: str | None = None
    fileHash: str | None = None
    createdDate: str | None = None
    fileSizeBytes: int | None = None
    sourceElementIds: list[str] = Field(default_factory=list)


class AdminRagChunkDetailResponse(BaseModel):
    chunk: AdminRagChunkDetail
    activeBuildId: str

File: frontend/lib/adminKnowledgeBase.ts
const API_BASE = (
  process.env.NEXT_PUBLIC_API_URL ??
  process.env.NEXT_PUBLIC_API_BASE_URL ??
  "http://127.0.0.1:8000"
).replace(/\/$/, "");

export class KnowledgeBaseApiError extends Error {
  status: number;

  constructor(message: string, status: number) {
    super(message);
    this.name = "KnowledgeBaseApiError";
    this.status = status;
  }
}

function authHeaders(): HeadersInit {
  if (typeof window === "undefined") return {};

  const token =
    window.localStorage.getItem("internova_access_token") ||
    window.localStorage.getItem("access_token") ||
    window.localStorage.getItem("token") ||
    window.localStorage.getItem("auth_token");

  return token ? { Authorization: `Bearer ${token}` } : {};
}

async function request<T>(path: string, init?: RequestInit): Promise<T> {
  const isFormData = init?.body instanceof FormData;
  const response = await fetch(`${API_BASE}${path}`, {
    ...init,
    cache: "no-store",
    headers: {
      Accept: "application/json",
      ...(isFormData ? {} : { "Content-Type": "application/json" }),
      ...authHeaders(),
      ...(init?.headers ?? {}),
    },
  });

  if (!response.ok) {
    let message = `HTTP ${response.status}`;

    try {
      const body = await response.json();
      if (typeof body?.detail === "string") {
        message = body.detail;
      }
    } catch {}

    throw new KnowledgeBaseApiError(message, response.status);
  }

  return response.json() as Promise<T>;
}

export type KnowledgeDocumentType =
  | "PDF"
  | "DOCX";

export type KnowledgeRagDocumentType =
  | "policy"
  | "form"
  | "agreement"
  | "talent_handbook"
  | "capstone_booklet"
  | "knowledge";

export type KnowledgeDocumentVersion = {
  id: number;
  version: string;
  fileUrl: string | null;
  fileHash: string | null;
  extractedTextPath: string | null;
  chunkPath: string | null;
  effectiveDate: string | null;
  status: string;
  createdAt: string | null;
};

export type KnowledgeIndexJob = {
  id: number;
  jobType: string;
  status: string;
  chunksCreated: number;
  errorMessage: string | null;
  startedAt: string | null;
  completedAt: string | null;
  createdAt: string | null;
};

export type KnowledgeIndexStatus = {
  status: "READY" | "DEGRADED" | "NOT_READY" | string;
  activeBuildId: string | null;
  documentsIndexed: number;
  chunksIndexed: number;
  activatedAtUnix: number | null;
  activatedAt: string | null;
  pointerExists: boolean;
  chromaReady: boolean;
  bm25Ready: boolean;
  manifestReady: boolean;
  pointerError: string | null;
  lastJobStatus: string | null;
  lastJobStartedAt: string | null;
  lastJobCompletedAt: string | null;
  lastJobError: string | null;
};

export type KnowledgeReindexResponse = {
  buildId: string;
  status: string;
  documentsIndexed: number;
  chunksCreated: number;
  sourceDir: string;
  ragDir: string;
  chromaDir: string;
  bm25Path: string;
  manifestPath: string;
  documents: string[];
  durationSeconds: number;
  removedBuilds: string[];
};

export type KnowledgeReloadResponse = {
  status: string;
  message: string;
};

export type KnowledgeUser = {
  id: number;
  fullName: string;
  email: string;
};

export type KnowledgeDocument = {
  id: number;
  title: string;
  documentType: KnowledgeDocumentType;
  ragDocumentType: KnowledgeRagDocumentType | null;
  description: string | null;
  fileUrl: string | null;
  currentVersion: string | null;
  year: number | null;
  status: string;
  uploadedBy: KnowledgeUser | null;
  createdAt: string | null;
  updatedAt: string | null;
  currentVersionInfo: KnowledgeDocumentVersion | null;
  latestIndexJob: KnowledgeIndexJob | null;
};

export type KnowledgeDocumentDetail = KnowledgeDocument & {
  versions: KnowledgeDocumentVersion[];
};

export type KnowledgeDocumentFilters = {
  types: string[];
  statuses: string[];
  years: number[];
};

export type KnowledgeDocumentsParams = {
  search?: string;
  documentType?: string;
  status?: string;
  year?: number;
  page?: number;
  pageSize?: number;
};

export type KnowledgeDocumentsResponse = {
  items: KnowledgeDocument[];
  total: number;
  page: number;
  pageSize: number;
  totalPages: number;
  filters: KnowledgeDocumentFilters;
};

export type KnowledgeDocumentDetailResponse = {
  document: KnowledgeDocumentDetail;
};

export type KnowledgeDocumentPayload = {
  title: string;
  documentType: KnowledgeDocumentType;
  ragDocumentType: KnowledgeRagDocumentType;
  description: string | null;
  fileUrl: string | null;
  currentVersion: string | null;
  year: number | null;
  status: string;
};

export type KnowledgeDocumentActionResponse = {
  documentId: number;
  message: string;
};

export type KnowledgeDocumentVersionsResponse = {
  items: KnowledgeDocumentVersion[];
};

export type KnowledgeVersionActionResponse = {
  documentId: number;
  versionId: number;
  message: string;
};

export type KnowledgeVersionUploadPayload = {
  version: string;
  status: string;
  effectiveDate: string | null;
  file: File;
};

export type KnowledgeChunkSummary = {
  total: number;
  documents: number;
  translated: number;
  averageCharacters: number;
};

export type KnowledgeChunkFilters = {
  documentNames: string[];
  documentTypes: string[];
  languages: string[];
};

export type KnowledgeChunkListItem = {
  chunkId: string;
  position: number;
  documentName: string;
  documentType: string;
  sourcePriority: number;
  contentPreview: string;
  language: string;
  page: number | null;
  section: string | null;
  subsection: string | null;
  topic: string | null;
  policyVersion: string | null;
  effectiveDate: string | null;
  ingestedAt: string | null;
  characterCount: number;
  wordCount: number;
  sourceElementCount: number;
  hasTranslation: boolean;
};

export type KnowledgeChunksParams = {
  search?: string;
  documentName?: string;
  documentType?: string;
  language?: string;
  page?: number;
  pageSize?: number;
};

export type KnowledgeChunksResponse = {
  items: KnowledgeChunkListItem[];
  summary: KnowledgeChunkSummary;
  filters: KnowledgeChunkFilters;
  activeBuildId: string;
  page: number;
  pageSize: number;
  total: number;
  totalPages: number;
};

export type KnowledgeChunkDetail = KnowledgeChunkListItem & {
  contentOriginal: string;
  contentVi: string | null;
  fileHash: string | null;
  createdDate: string | null;
  fileSizeBytes: number | null;
  sourceElementIds: string[];
};

export type KnowledgeChunkDetailResponse = {
  chunk: KnowledgeChunkDetail;
  activeBuildId: string;
};

function documentsQuery(params: KnowledgeDocumentsParams): string {
  const query = new URLSearchParams();

  if (params.search?.trim()) query.set("search", params.search.trim());
  if (params.documentType) query.set("document_type", params.documentType);
  if (params.status) query.set("status", params.status);
  if (params.year) query.set("year", String(params.year));
  query.set("page", String(params.page ?? 1));
  query.set("page_size", String(params.pageSize ?? 10));

  return query.toString();
}

function chunksQuery(params: KnowledgeChunksParams): string {
  const query = new URLSearchParams();
  if (params.search?.trim()) query.set("search", params.search.trim());
  if (params.documentName) query.set("document_name", params.documentName);
  if (params.documentType) query.set("document_type", params.documentType);
  if (params.language) query.set("language", params.language);
  query.set("page", String(params.page ?? 1));
  query.set("page_size", String(params.pageSize ?? 25));
  return query.toString();
}

export const adminKnowledgeBaseApi = {

    indexStatus: () =>
    request<KnowledgeIndexStatus>(
      "/api/v1/admin/knowledge/index-status",
    ),

  reindex: () =>
    request<KnowledgeReindexResponse>(
      "/api/v1/admin/knowledge/reindex",
      {
        method: "POST",
      },
    ),

  chunks: (params: KnowledgeChunksParams = {}) =>
    request<KnowledgeChunksResponse>(
      `/api/v1/admin/knowledge/chunks?${chunksQuery(params)}`,
    ),

  chunk: (chunkId: string) =>
    request<KnowledgeChunkDetailResponse>(
      `/api/v1/admin/knowledge/chunks/${encodeURIComponent(chunkId)}`,
    ),

  documents: (params: KnowledgeDocumentsParams = {}) =>
    request<KnowledgeDocumentsResponse>(
      `/api/v1/admin/knowledge/documents?${documentsQuery(params)}`,
    ),

  document: (id: number) =>
    request<KnowledgeDocumentDetailResponse>(
      `/api/v1/admin/knowledge/documents/${encodeURIComponent(id)}`,
    ),

  createDocument: (payload: KnowledgeDocumentPayload) =>
    request<KnowledgeDocumentDetailResponse>(
      "/api/v1/admin/knowledge/documents",
      {
        method: "POST",
        body: JSON.stringify(payload),
      },
    ),

  updateDocument: (id: number, payload: Partial<KnowledgeDocumentPayload>) =>
    request<KnowledgeDocumentDetailResponse>(
      `/api/v1/admin/knowledge/documents/${encodeURIComponent(id)}`,
      {
        method: "PATCH",
        body: JSON.stringify(payload),
      },
    ),

  archiveDocument: (id: number) =>
    request<KnowledgeDocumentDetailResponse>(
      `/api/v1/admin/knowledge/documents/${encodeURIComponent(id)}/archive`,
      {
        method: "POST",
      },
    ),

  deleteDocument: (id: number) =>
    request<KnowledgeDocumentActionResponse>(
      `/api/v1/admin/knowledge/documents/${encodeURIComponent(id)}`,
      {
        method: "DELETE",
      },
    ),

  versions: (documentId: number) =>
    request<KnowledgeDocumentVersionsResponse>(
      `/api/v1/admin/knowledge/documents/${encodeURIComponent(documentId)}/versions`,
    ),

  uploadVersion: (
    documentId: number,
    payload: KnowledgeVersionUploadPayload,
  ) => {
    const body = new FormData();
    body.set("version", payload.version);
    body.set("status", payload.status);
    if (payload.effectiveDate) {
      body.set("effective_date", payload.effectiveDate);
    }
    body.set("file", payload.file);

    return request<KnowledgeVersionActionResponse>(
      `/api/v1/admin/knowledge/documents/${encodeURIComponent(documentId)}/versions`,
      {
        method: "POST",
        body,
      },
    );
  },

  setCurrentVersion: (documentId: number, versionId: number) =>
    request<KnowledgeVersionActionResponse>(
      `/api/v1/admin/knowledge/documents/${encodeURIComponent(documentId)}/versions/${encodeURIComponent(versionId)}/set-current`,
      {
        method: "POST",
      },
    ),

  reload: () =>
    request<KnowledgeReloadResponse>(
      "/api/v1/chat/reload",
      {
        method: "POST",
      },
    ),
};

export async function openKnowledgeDocumentVersion(
  documentId: number,
  versionId: number,
  download = false,
): Promise<void> {
  const response = await fetch(
    `${API_BASE}/api/v1/admin/knowledge/documents/${encodeURIComponent(documentId)}/versions/${encodeURIComponent(versionId)}/file${download ? "?download=true" : ""}`,
    { headers: { ...authHeaders() } },
  );
  if (!response.ok) {
    let message = `HTTP ${response.status}`;
    try {
      const body = (await response.json()) as { detail?: string };
      if (body.detail) message = body.detail;
    } catch {}
    throw new KnowledgeBaseApiError(message, response.status);
  }

  const blobUrl = URL.createObjectURL(await response.blob());
  if (download) {
    const link = document.createElement("a");
    link.href = blobUrl;
    link.download = "";
    link.click();
  } else {
    window.open(blobUrl, "_blank", "noopener,noreferrer");
  }
  window.setTimeout(() => URL.revokeObjectURL(blobUrl), 60_000);
}

export function formatDateTime(value: string | null | undefined): string {
  if (!value) return "-";

  const date = new Date(value);
  if (Number.isNaN(date.getTime())) return value;

  return new Intl.DateTimeFormat("vi-VN", {
    year: "numeric",
    month: "2-digit",
    day: "2-digit",
    hour: "2-digit",
    minute: "2-digit",
  }).format(date);
}
